### PR TITLE
Test: added additional tests for GenreHierarchy

### PR DESCRIPTION
**Pull request by etimr**

### CODE REVIEW

File: srv/src/test/java/my/bookshop/handlers/HierarchyHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		h1.setName("Philosophical fiction");
+		GenreHierarchy h2 = GenreHierarchy.create();
+		h2.setName("Epic");
+		List<GenreHierarchy> list = List.of(h1,h2);
+		List<GenreHierarchy> sorted = list.stream().sorted(new Sorter()).toList();
+		assertEquals("Epic", sorted.get(0).getName());
+	}
+
+	@Test
+	public void testSortingChildren() {
```
Comment (eugene-andreev) [inline srv/src/test/java/my/bookshop/handlers/HierarchyHandlerTest.java:24]:
```suggestion
	public void testSortingChildrenSameRoot() {
```
or
```suggestion
	public void testSorting`Siblings() {
```

File: srv/src/test/java/my/bookshop/handlers/HierarchyHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	public void testSortingChildrenSameRoot() {
+		GenreHierarchy root = GenreHierarchy.create();
+		root.setName("Folklore");
+		GenreHierarchy h1 = GenreHierarchy.create();
+		h1.setName("Urban legend");
+		h1.setParnt(root);
+		List<GenreHierarchy> list = List.of(h1,root);
+		List<GenreHierarchy> sorted = list.stream().sorted(new Sorter()).toList();
+		assertEquals("Folklore", sorted.get(0).getName());
+	}
```
Comment (eugene-andreev) [inline srv/src/test/java/my/bookshop/handlers/HierarchyHandlerTest.java:62]:
```suggestion
	public void testSortingChildrenSameRoot() {
		GenreHierarchy root = hierarchy("Folklore");
		GenreHierarchy h1 = hierarchy("Urban legend", root);
		List<GenreHierarchy> sorted = sorted(h1, root);
		assertEquals("Folklore", sorted.get(0).getName());
	}

	private static GenreHierarchy hierarchy(String name, GenreHierarchy parent) {
		GenreHierarchy h1 = GenreHierarchy.create();
		h1.setName(name);
		if (parent != null) {
			h1.setParnt(parent);
		}
		return h1;
	}

	private static GenreHierarchy hierarchy(String name) {
		return hierarchy(name, null);
	}

	private static List<GenreHierarchy> sorted(GenreHierarchy... h) {
		List<GenreHierarchy> list = List.of(h);
		return list.stream().sorted(new Sorter()).toList();
	}
```

File: srv/src/test/java/my/bookshop/handlers/HierarchyHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		root.setName("Folklore");
+		GenreHierarchy h1 = GenreHierarchy.create();
+		h1.setName("Urban legend");
+		h1.setParnt(root);
+		GenreHierarchy h2 = GenreHierarchy.create();
+		h2.setName("Fairy tale");
+		h2.setParnt(root);
+		List<GenreHierarchy> list = List.of(h1,h2);
+		List<GenreHierarchy> sorted = list.stream().sorted(new Sorter()).toList();
+		assertEquals("Fairy tale", sorted.get(0).getName());
```
Comment (agoerler) [inline srv/src/test/java/my/bookshop/handlers/HierarchyHandlerTest.java:35]: please also check the positions of "Folklore" and "Urban Legend"

File: srv/src/test/java/my/bookshop/handlers/HierarchyHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		GenreHierarchy root1 = GenreHierarchy.create();
+		root1.setName("Thriller");
+		GenreHierarchy root2 = GenreHierarchy.create();
+		root2.setName("Folklore");
+		GenreHierarchy h2 = GenreHierarchy.create();
+		h2.setName("Urban legend");
+		h2.setParnt(root2);
+		List<GenreHierarchy> list = List.of(root1,h2);
+		List<GenreHierarchy> sorted = list.stream().sorted(new Sorter()).toList();
+		assertEquals("Urban legend", sorted.get(0).getName());
```
Comment (agoerler) [inline srv/src/test/java/my/bookshop/handlers/HierarchyHandlerTest.java:49]: please also check the positions of "Thriller" and "Urban Legend"

File: srv/src/test/java/my/bookshop/GenreHierarchyTest.java [hunk 5]
```diff
@@ -94,18 +127,125 @@ void testCollapseAll() throws Exception {
...
+  			.andExpect(jsonPath("$.value[15]").doesNotExist());
+		}
+	}
+
+	@Test
+	@WithMockUser(username = "admin")
+	void testSearch() throws Exception {
+		assumeThat(env.getActiveProfiles()).doesNotContain("hybrid");
 		client.perform(get(genresURI
 				+ "?$select=DistanceFromRoot,DrillState,ID,LimitedDescendantCount,name"
```
[the remark this answers is not in the document]
Comment (etimr) [reply srv/src/test/java/my/bookshop/GenreHierarchyTest.java:165]: we need to fix it on HANA first, there is a bug with orderBy and siblings -> BLI created

File: app/router/xs-app.json [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      "cacheControl": "no-cache, no-store, must-revalidate"
+    },
+    {
+      "source": "^/(.*)$",
+      "target": "$1",
+      "destination": "srv-api",
+      "csrfProtection": true
+    }
+  ]
+}
```
Comment (agoerler) [inline app/router/xs-app.json:27]: no idea where these changes came from :-(

File: srv/src/test/java/my/bookshop/GenreHierarchyTest.java [hunk 7]
```diff
@@ -113,8 +140,108 @@ void testExpandAll() throws Exception {
...
+				+ "/com.sap.vocabularies.Hierarchy.v1.TopLevels(HierarchyNodes=$root/GenreHierarchy,HierarchyQualifier='GenreHierarchy',NodeProperty='ID',Levels=1)"))
+				.andExpect(status().isOk())
+				.andExpect(jsonPath("$.value[0].ID").value(20))
+				.andExpect(jsonPath("$.value[0].name").value("Non-Fiction"))
+				.andExpect(jsonPath("$.value[0].DrillState").value("collapsed"))
+				.andExpect(jsonPath("$.value[0].DistanceFromRoot").value(0))
+				.andExpect(jsonPath("$.value[1]").doesNotExist());
+	}
+
+	@Test
```
Comment (eugene-andreev) [inline srv/src/test/java/my/bookshop/GenreHierarchyTest.java:198]: This test currently is red, when running on HANA. The fix will come only in 3.6.1.
I would ignore it so far. What do you think?

Comment (eugene-andreev) [inline srv/src/test/java/my/bookshop/GenreHierarchyTest.java:198]:
```suggestion
	@Ignore
	@Test
```

File: srv/src/test/java/my/bookshop/GenreHierarchyTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (etimr) [reply srv/src/test/java/my/bookshop/GenreHierarchyTest.java:152]: Evgeny and I checked the test, it's green both locally and on HANA. Evgeny added the check for `LimitedDescendantCount`

File: srv/src/test/java/my/bookshop/GenreHierarchyTest.java (hunk 7 above)
Comment (agoerler) [reply srv/src/test/java/my/bookshop/GenreHierarchyTest.java:198]: fine with me

**Review by agoerler — APPROVED** [review]

**Pull request merged**